### PR TITLE
[typing] prefect.client

### PR DESCRIPTION
**Pull request by mjpieters**
Code now passes pyright checking in strict mode.

### CODE REVIEW

File: src/prefect/main.py [hunk 1]
```diff
@@ -1,4 +1,7 @@
 # Import user-facing API
+from typing import Any
+
+from pluggy import Result
```
Comment (desertaxle) [inline src/prefect/main.py:4]: I think this came in by accident and is causing the failures in CI.

Comment (mjpieters) [reply src/prefect/main.py:4]: Yup, just found it too. Not sure how my IDE managed this.

**Review by desertaxle — APPROVED** [review]
Thanks @mjpieters!

**Review by zzstoatzz — APPROVED** [review]

**Comment by zzstoatzz** [comment]
thank you @mjpieters !

File: src/prefect/client/utilities.py [hunk 2]
```diff
@@ -5,32 +5,31 @@
...
+R = TypeVar("R", infer_variance=True)
+
+
+def _current_async_client(
+    client: Union["PrefectClient", "SyncPrefectClient"],
+) -> TypeIs["PrefectClient"]:
+    from prefect._internal.concurrency.event_loop import get_running_loop
+
+    # Only a PrefectClient will have a _loop attribute that is the current loop
+    return getattr(client, "_loop", None) == get_running_loop()
```
[the remark this answers is not in the document]
Comment (mjpieters) [reply src/prefect/client/utilities.py:27]: I may have used the wrong special form here. This type narrowing function returns a `TypeIs` construct but now that I reread this it should have been a `TypeGuard` instead.

That's because when this function returns False in two different cases:

1. the client is a `SynPrefectClient` instance.
2. the client is a `PrefectClient` instance, attached to a different loop.

The `TypeIs` construct acts like a `isinstance()` check, where the type checker can infer that if False is returned then the client must be a `SyncPrefectClient`. But that's not correct for case 2.

A `TypeGuard` function explicitly doesn't make promises about types when False is returned and so fits this function much better.

It is a technically moot point **in how this function is used in this module** but an important distinction if you were to use this as an example 🤦.

TLDR: 

- use `TypeIs[R]` when True means _this is of type `R`_ and False means _this is not of type `R`_
- use `TypeGuard[R]` when True means _this meets conditions that tell me this is of type `R`_ but False means _uh, I am not sure what this is and I don't care 🤷_.
